granulate_utils/containers/cri.py

```diff
@@ -70,29 +71,49 @@ def _reconstruct_name(self, container: api_pb2.Container) -> str:
     def list_containers(self, all_info: bool) -> List[Container]:
         containers: List[Container] = []
 
-        with RuntimeServiceWrapper(self._path) as stub:
-            for container in stub.ListContainers(api_pb2.ListContainersRequest()).containers:
-                if all_info:
-                    # need verbose=True to get the info which contains the PID
+        for rt, path in self._runtimes.items():
+            with RuntimeServiceWrapper(path) as stub:
+                for container in stub.ListContainers(api_pb2.ListContainersRequest()).containers:
+                    if all_info:
+                        # need verbose=True to get the info which contains the PID
+                        status = stub.ContainerStatus(
+                            api_pb2.ContainerStatusRequest(container_id=container.id, verbose=True)
+                        )
+                        pid: Optional[int] = json.loads(status.info.get("info", "{}")).get("pid")
+                    else:
+                        pid = None
+
+                    containers.append(self._create_container(container, pid, rt))
+
+        return containers
+
+    def get_container(self, container_id: str, all_info: bool) -> Container:
+        for rt, path in self._runtimes.items():
+            with RuntimeServiceWrapper(path) as stub:
+                try:
                     status = stub.ContainerStatus(
-                        api_pb2.ContainerStatusRequest(container_id=container.id, verbose=True)
-                    )
-                    pid: Optional[int] = json.loads(status.info.get("info", "{}")).get("pid")
-                else:
-                    pid = None
-
-                containers.append(
-                    Container(
-                        runtime=self._runtime_name,
-                        name=self._reconstruct_name(container),
-                        id=container.id,
-                        labels=container.labels,
-                        running=container.state == CONTAINER_RUNNING,
-                        pid=pid,
+                        api_pb2.ContainerStatusRequest(container_id=container_id, verbose=all_info)
                     )
-                )
+                except grpc._channel._InactiveRpcError:
```

Comment:
Why `except InactiveRpcError` here? (and if here, why not in `list_containers`?)

Reply:
And what if the container does not exist under containerd but does exist under cri-o?

Reply:
You answered you own question :)
If the container doesn't exist under containerd it raises an `_InactiveRpcError` (A bit of a misleading name I think) with status = `StatusCode.NOT_FOUND`, and then we `continue` to try cri-o.

```
_InactiveRpcError: <_InactiveRpcError of RPC that terminated with:
	status = StatusCode.NOT_FOUND
	details = "an error occurred when try to find container "asdf": not found"
	debug_error_string = "{"created":"@1646903927.891701180","description":"Error received from peer unix:/run/containerd/containerd.sock","file":"src/core/lib/surface/call.cc","file_line":903,"grpc_message":"an error occurred when try to find container "asdf": not found","grpc_status":5}"
>

_InactiveRpcError: <_InactiveRpcError of RPC that terminated with:
	status = StatusCode.NOT_FOUND
	details = "could not find container "asdf": container with ID starting with asdf not found: ID does not exist"
	debug_error_string = "{"created":"@1646903802.171836819","description":"Error received from peer unix:/var/run/crio/crio.sock","file":"src/core/lib/surface/call.cc","file_line":903,"grpc_message":"could not find container "asdf": container with ID starting with asdf not found: ID does not exist","grpc_status":5}"
>
```

Reply:
In `list_containers` we don't need it because we assume the connection is valid (since we checked it in the ctor) and there's no other reason for list_containers to fail

Reply:
I see. Indeed misleading, I knew that `InactiveRpcError` happens when the socket is closed. Can we filter the type and not continue if it's something else? i.e, `continue` only if `status == StatusCode.NOT_FOUND`?

Reply:
Fixed